Source/DafnyLanguageServer/Plugins/DafnyCodeActionHelpers.cs

```diff
@@ -0,0 +1,193 @@
+    var pos = endToken.pos + endToken.val.Length - 1;
+    var indentation = 0;
+    var indentationBrace = endToken.col - 1;
+    var firstNewline = true;
+    var useTabs = false;
+    // Look for the first newline
+    while (0 <= pos && pos < text.Length && (text[pos] != '\n' || firstNewline)) {
+      if (text[pos] == '\t') {
+        useTabs = true;
+      }
+
+      if (text[pos] == '\n') {
+        if (!firstNewline) {
+          break;
+        }
+
+        firstNewline = false;
+      }
+
+      if (!firstNewline) {
+        if (text[pos] == ' ' || text[pos] == '\t') {
+          indentation++;
+        } else {
+          indentation = 0;
+        }
+      }
+
+      pos--;
+    }
+
+    if (startLine == endToken.line - 1) {
+      // Override case {\n} with some spacing, we return a default value of 2 spaces or 1 tab, if we find some
+      indentation = indentationBrace + (useTabs ? 1 : 2);
+    } else if (startLine == endToken.line) {
+      // Override case { ... } on one line.
+      indentationBrace = startCol - 1;
+      indentation = indentationBrace + 1;
+    }
+
+    indentation = Math.Max(indentationBrace, indentation);
+
+    return (
+      new string(useTabs ? '\t' : ' ', indentation - indentationBrace),
+      new string(useTabs ? '\t' : ' ', indentationBrace));
+  }
+
+  /// <summary>
+  /// Given the position of an opening brace of a Block, returns the range for the position just before the closing brace,
+  /// and indentation helpers as defined by GetIndentationBefore()
+  /// </summary>
+  /// <param name="input"></param>
+  /// <param name="openingBracePosition"></param>
+  /// <returns></returns>
+  public static (Range? beforeEndBrace, string indentationExtra, string indentationUntilBrace)
+      GetInformationToInsertAtEndOfBlock(IDafnyCodeActionInput input, Position openingBracePosition) {
+
+    var (line, col) = openingBracePosition.ToTokenLineAndCol();
+    var endToken = GetMatchingEndToken(input.Program!, input.Uri, line, col);
+    if (endToken == null) {
+      return (null, "", "");
+    }
+
+    var (extraIndentation, indentationUntilBrace) = GetIndentationBefore(endToken, line, col, input.Code);
+    var beforeClosingBrace = endToken.GetLspRange().GetStartRange();
+    return (beforeClosingBrace, extraIndentation, indentationUntilBrace);
+  }
+
+  /// <summary>
+  /// Given an opening brace of a Block (typically where an error is detected),
+  /// finds and returns the closing brace token.
+  /// </summary>
+  /// <param name="program">The program</param>
+  /// <param name="documentUri">The document URI</param>
+  /// <param name="line">The line of the opening brace</param>
+  /// <param name="col">The column of the opening brace</param>
+  /// <returns>The token of a matching closing brace, typically the `ÈndTok` of a BlockStmt</returns>
+  private static IToken? GetMatchingEndToken(Dafny.Program program, string documentUri, int line, int col) {
```

Comment:
Why not pass a `Position openingBrace` instead of a line and col ?

Reply:
Because we need to compare the line and the column to a lot of tokens, and Position is offset by tokens by 1, so we would be computing unnecessary affine transformations all the time.
This is a private method anyway, we can always upgrade this later.

Reply:
This seems like a slow approach to find the end token. I think something like the IIntervalTree used [here](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyLanguageServer/Language/Symbols/SymbolTable.cs#L33), or a position directed search like [here](https://github.com/keyboardDrummer/Miksilo/blob/f04b13466426d5cbdd7766f4c8c6239f7241a675/languageServer/shared/src/main/scala/miksilo/languageServer/server/SourcePath.scala#L19) would be better since both are constant instead of linear complexity.

Reply:
I agree we should do a better job at indexing the program to access various definitions, as we get more and more features about navigation.
For this PR, implementing a tree system is out of scope, because the verification completion plugin is a toy example and is meant to evolve; to avoid traversing the body of 10k methods, I added a check about line numbers, so that it will be very fast. I think that's the best reasonable we should aim at for now if we want this PR to be merged.